cmd/logger.go

```diff
@@ -154,36 +161,42 @@ func logIf(level logrus.Level, source string, err error, msg string, data ...int
 	if err == nil || isErrIgnored(err) {
 		return
 	}
-
-	fields := logrus.Fields{
-		"source": source,
-		"cause":  err.Error(),
-	}
-
-	if terr, ok := err.(*errors.Error); ok {
-		fields["stack"] = strings.Join(terr.Stack(), " ")
+	var stack []string
+	cause := err.Error()
+	// Get stack trace
+	stack = trace(3)
+	// Get time
+	time := time.Now().UTC().Format(time.RFC3339Nano)
+	// Print formatted log message on the console
+	stackTab := "       "
+	levelCaps := strings.ToUpper(level)
+	message := fmt.Sprintf(msg, data...)
+	if log.json {
+		logJSON, _ := json.Marshal(&logEntry{
+			Level:   levelCaps,
+			Message: message,
+			Time:    time,
+			Cause:   cause,
+			Source:  source,
+			Stack:   stack,
+		})
+		fmt.Println(string(logJSON))
+	} else {
+		console.Printf("Trace: %s\n", strings.Join(reverse(stack), "\n"+stackTab))
+		console.Printf("Source: %s\n", source)
+		console.Printf("Time: %s\n", time)
+		console.Printf("Cause: %s\n", cause)
+		console.Printf(colorRed(colorBold(levelCaps + ": " + message + "\n\n")))
```

Comment:
1. Similar to `json` marshaling, create an output string and print after `if`
2. `Time` in the middle looks strange to me.  I see human readable log has time first.  We have to revisit order of elements in the log
3. Don't know why we need to have labels for values i.e. `Time: time`

My choice is like below
```
<Time> <Source> <Message> <Cause> Stack:<Trace>
```

Reply:
1. Created a new string variable, `logRegular`, using `fmt.Sprintf` command instead of printing each item one by one with `console.Printf` :
```
    :
    fmt.Println(string(logJSON))
} else {
    logRegular := fmt.Sprintf("Trace: %s\n", strings.Join(stack, "\n"+stackTab)) +
	fmt.Sprintf("Source: %s\n", source) +
	fmt.Sprintf("Time: %s\n", time) +
	fmt.Sprintf("Cause: %s\n", cause) +
	colorRed(colorBold(level.String()+": "+message+"\n\n"))
    fmt.Println(logRegular)
}
```
I preferred keeping the print commands inside `if/else` blocks to make it a little bit more clear by the help of the argument names; `if` part prints `logJSON`  and `else` part prints `logRegular`.

2. & 3. We thought, it would be useful to know when and how frequent an issue happens time-wise. Additionally, there is a plan to have a `log message grouping` feature in near future to collect identical log messages in a pool and show them only once if they happen in a certain time interval/window.
We also decided during out discussions, it'll be more clear/readable if we don't have each log field information chained back to back in a single line.

Please see the following 2 examples:

When chained (old way),
```
ERRO[0032] Unable to validate content-md5 format.        cause=illegal base64 data at
input byte4 source=[object-handlers.go:491:objectAPIHandlers.PutObjectHandler()]
ERRO[0034] {"method":"PUT","reqURI":"/awscli-mint-test-bucket-2279/datafile-1-MB","header":
{"Accept-Encoding":["identity"],"Authorization":["AWS4-HMAC-SHA256
Credential=minio/20171229/us-east-1/s3/aws4_request, SignedHeaders=content-length;
content-md5;host;x-amz-content-sha256;x-amz-date,
Signature=389853a9186cdea3f387ef708e729dce147c461c1bf3ee91572f31eb4eaa5cb2"],
"Content-Length":["1048576"],"Content-Md5":["YF2PEiZtBMmzbzWNF4O6Hg=="],"Expect":
["100-continue"],"Host":["127.0.0.1:9000"],"User-Agent":["aws-cli/1.11.112 Python/3.5.2
Linux/4.13.0-21-generic botocore/1.5.75"],"X-Amz-Content-Sha256":
["ebba76af7ddcba18c52f567ea226255b9a4ae15bda9de0c0b78bd998d17237cd"],"X-Amz-Date":
["20171229T003707Z"]}}  cause=Signature does not match source=[object-
handlers.go:590:objectAPIHandlers.PutObjectHandler()]
```
versus in separate  lines with labels (new approach),
```
Trace: /home/ersan/go/src/runtime/asm_amd64.s:2337:runtime.goexit
       /home/ersan/go/src/net/http/server.go:1801:http.(*conn).serve
       /home/ersan/go/src/net/http/server.go:2619:http.serverHandler.ServeHTTP
       /home/ersan/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP
       /home/ersan/work/src/github.com/minio/minio/pkg/http/server.go:107:http.(*Server).Start.func1
       <autogenerated>:1:cmd.(*reservedMetadataHandler).ServeHTTP
       generic-handlers.go:133:cmd.reservedMetadataHandler.ServeHTTP
       <autogenerated>:1:cmd.(*authHandler).ServeHTTP
       auth-handler.go:234:cmd.authHandler.ServeHTTP
       <autogenerated>:1:cmd.(*resourceHandler).ServeHTTP
       generic-handlers.go:448:cmd.resourceHandler.ServeHTTP
       /home/ersan/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP
       /home/ersan/work/src/github.com/minio/minio/vendor/github.com/rs/cors/cors.go:190:cors.(*Cors).Handler.func1
       <autogenerated>:1:cmd.(*timeValidityHandler).ServeHTTP
       generic-handlers.go:337:cmd.timeValidityHandler.ServeHTTP
       <autogenerated>:1:cmd.(*cacheControlHandler).ServeHTTP
       generic-handlers.go:246:cmd.cacheControlHandler.ServeHTTP
       <autogenerated>:1:cmd.(*minioReservedBucketHandler).ServeHTTP
       generic-handlers.go:267:cmd.minioReservedBucketHandler.ServeHTTP
       <autogenerated>:1:cmd.(*redirectHandler).ServeHTTP
       generic-handlers.go:219:cmd.redirectHandler.ServeHTTP
       <autogenerated>:1:cmd.(*crossDomainPolicy).ServeHTTP
       crossdomain-xml-handler.go:51:cmd.crossDomainPolicy.ServeHTTP
       <autogenerated>:1:cmd.(*requestHeaderSizeLimitHandler).ServeHTTP
       generic-handlers.go:90:cmd.requestHeaderSizeLimitHandler.ServeHTTP
       <autogenerated>:1:cmd.(*requestSizeLimitHandler).ServeHTTP
       generic-handlers.go:65:cmd.requestSizeLimitHandler.ServeHTTP
       <autogenerated>:1:cmd.(*httpStatsHandler).ServeHTTP
       generic-handlers.go:497:cmd.httpStatsHandler.ServeHTTP
       <autogenerated>:1:cmd.(*pathValidityHandler).ServeHTTP
       generic-handlers.go:558:cmd.pathValidityHandler.ServeHTTP
       <autogenerated>:1:cmd.(*rateLimit).ServeHTTP
       generic-handlers.go:600:cmd.rateLimit.ServeHTTP
       /home/ersan/work/src/github.com/minio/minio/vendor/github.com/gorilla/mux/mux.go:107:mux.(*Router).ServeHTTP
       /home/ersan/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP
       api-router.go:63:cmd.(objectAPIHandlers).PutObjectHandler-fm
       object-handlers.go:491:cmd.objectAPIHandlers.PutObjectHandler
Source: [object-handlers.go:491:objectAPIHandlers.PutObjectHandler()]
Time: 2017-12-29T00:33:16.239282515Z
Cause: illegal base64 data at input byte 4
ERROR: Unable to validate content-md5 format.
```

Reply:
For (1):
The primary reason is to move away `fmt.Println()` in `if` and `else` block.  Similarly why you need to have multiple `fmt.Sprint()` and adding them to one string.  Below is much cleaner
```go
if json {
...
output = string(data)
} else {
output = fmt.Sprintf(
`Trace: %s
Source: %s
Time: %s
Cause: %s
%s`, strings.Join(reverse(stack), "\n"+stackTab), source, time, cause, colorRed(colorBold(levelCaps + ": " + message + "\n\n")))
}
fmt.Println(output)
```

For rest:  Why I am suggesting `<Time> <Source> <Message> <Cause> Stack:<Trace>` is readability problem with field naming type.  In my suggestion, there is multi-liner stack trace, for the above same, it supposes to look like below
```
[2017-12-29T00:33:16.239282515Z] [object-handlers.go:491:objectAPIHandlers.PutObjectHandler()] ERROR: Unable to validate content-md5 format. illegal base64 data at input byte 4.
Trace: /home/ersan/go/src/runtime/asm_amd64.s:2337:runtime.goexit
       /home/ersan/go/src/net/http/server.go:1801:http.(*conn).serve
       /home/ersan/go/src/net/http/server.go:2619:http.serverHandler.ServeHTTP
       /home/ersan/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP
       /home/ersan/work/src/github.com/minio/minio/pkg/http/server.go:107:http.(*Server).Start.func1
       <autogenerated>:1:cmd.(*reservedMetadataHandler).ServeHTTP
       generic-handlers.go:133:cmd.reservedMetadataHandler.ServeHTTP
       <autogenerated>:1:cmd.(*authHandler).ServeHTTP
       auth-handler.go:234:cmd.authHandler.ServeHTTP
       <autogenerated>:1:cmd.(*resourceHandler).ServeHTTP
       generic-handlers.go:448:cmd.resourceHandler.ServeHTTP
       /home/ersan/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP
       /home/ersan/work/src/github.com/minio/minio/vendor/github.com/rs/cors/cors.go:190:cors.(*Cors).Handler.func1
       <autogenerated>:1:cmd.(*timeValidityHandler).ServeHTTP
       generic-handlers.go:337:cmd.timeValidityHandler.ServeHTTP
       <autogenerated>:1:cmd.(*cacheControlHandler).ServeHTTP
       generic-handlers.go:246:cmd.cacheControlHandler.ServeHTTP
       <autogenerated>:1:cmd.(*minioReservedBucketHandler).ServeHTTP
       generic-handlers.go:267:cmd.minioReservedBucketHandler.ServeHTTP
       <autogenerated>:1:cmd.(*redirectHandler).ServeHTTP
       generic-handlers.go:219:cmd.redirectHandler.ServeHTTP
       <autogenerated>:1:cmd.(*crossDomainPolicy).ServeHTTP
       crossdomain-xml-handler.go:51:cmd.crossDomainPolicy.ServeHTTP
       <autogenerated>:1:cmd.(*requestHeaderSizeLimitHandler).ServeHTTP
       generic-handlers.go:90:cmd.requestHeaderSizeLimitHandler.ServeHTTP
       <autogenerated>:1:cmd.(*requestSizeLimitHandler).ServeHTTP
       generic-handlers.go:65:cmd.requestSizeLimitHandler.ServeHTTP
       <autogenerated>:1:cmd.(*httpStatsHandler).ServeHTTP
       generic-handlers.go:497:cmd.httpStatsHandler.ServeHTTP
       <autogenerated>:1:cmd.(*pathValidityHandler).ServeHTTP
       generic-handlers.go:558:cmd.pathValidityHandler.ServeHTTP
       <autogenerated>:1:cmd.(*rateLimit).ServeHTTP
       generic-handlers.go:600:cmd.rateLimit.ServeHTTP
       /home/ersan/work/src/github.com/minio/minio/vendor/github.com/gorilla/mux/mux.go:107:mux.(*Router).ServeHTTP
       /home/ersan/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP
       api-router.go:63:cmd.(objectAPIHandlers).PutObjectHandler-fm
       object-handlers.go:491:cmd.objectAPIHandlers.PutObjectHandler
```

Reply:
1. Makes sense. Done.
2. We'll discuss your suggestion with AB.